src/librustc/mir/mod.rs

```diff
@@ -1336,6 +1383,10 @@ impl<'tcx> BasicBlockData<'tcx> {
         BasicBlockData { statements: vec![], terminator, is_cleanup: false }
     }
 
+    pub fn terminator_opt(&self) -> &Option<Terminator<'tcx>> {
+        &self.terminator
```

Comment:
Since I can't leave a comment outside of my changes...

I really wanted to make `BasicBlockData::terminator` private, but that's not possible because `Cfg` needs it in order to build up the `BasicBlockData` initially. Would it make sense to make `terminator` private and add a new method called something like `terminator_for_cfg` and `terminator_for_cfg_mut` that's intended only to be used by `Cfg`? This would help ensure that all other mutable access to `terminator` properly goes through `Body` so the cache can be properly maintained. In the current form, it would be very easy for someone to access `&mut terminator` directly without realizing it.

Reply:
yea, seems fine to makeit private and create a `BasicBlockData::new` function that takes all values needed to construct a `BasicBlockData`